github-api-call-inventory/github-api-call-inventory.ipynb

```diff
@@ -0,0 +1,765 @@
+    "\n",
+    "The only time it makes sense to load this from GitHub is when the pipeline is started, for example when the PR event is received."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "616bb63c",
+   "metadata": {},
+   "source": [
+    "Let's pick the worst case from the list above (*6 chroots, config is called 'packit.yaml'*) as a reference, and check how the number of API calls change if we apply the above.\n",
+    "\n",
+    "**Load the package config only once per pipeline run** to increase the number of possible runs from 11 to 24 (~118% increase). This is achived by reducing the number of API calls to get the package config from 222 to 6."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "c1b001e5",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "prs_per_hour_per_org = 24\n",
+       "total_is_project_private = 37\n",
+       "total_is_user_approved = 74\n",
+       "total_load_packit_yaml = 6\n",
+       "total_set_status = 84\n",
+       "grand_total = 201"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "github_api_calls(\n",
+    "    chroots=6, \n",
+    "    rate_limit=5000,\n",
+    "    load_packit_yaml=5+1,\n",
+    "    builds_enabled=True,\n",
+    "    tests_enabled=True,\n",
+    "    load_packit_yaml_once=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4808dd68",
+   "metadata": {},
+   "source": [
+    "**Check if the user is allowed to run the pipeline only once**, when the PR is updated. This helps getting rid of 72 more API calls, and get to 38 PRs/hour/org (~245% increase)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "38ba0b4f",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "prs_per_hour_per_org = 38\n",
+       "total_is_project_private = 37\n",
+       "total_is_user_approved = 2\n",
+       "total_load_packit_yaml = 6\n",
+       "total_set_status = 84\n",
+       "grand_total = 129"
+      ]
+     },
+     "execution_count": 12,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "github_api_calls(\n",
+    "    chroots=6, \n",
+    "    rate_limit=5000,\n",
+    "    load_packit_yaml=5+1,\n",
+    "    builds_enabled=True,\n",
+    "    tests_enabled=True,\n",
+    "    load_packit_yaml_once=True,\n",
+    "    user_check_once=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2284d964",
+   "metadata": {},
+   "source": [
+    "**Don't update the test flags while the build is in progress**, to get to 47 PRs/hour/org (a 327% increase in total), thanks to shaving of 24 API calls to update the status flags."
```

Comment:
I'm not sure about the UX here, but this could somehow help with the situation with GitLab statuses where we cannot overwrite "running" -> "running" status (packit/packit-service#1821): this way, we could have only "In progress" status pointing to the build/test correctly

Reply:
If GitLab has a "pending" status, and supports updates to that, then yes, this could help. Then the "running" status would be set only when the test is actually triggered.

Reply:
yes, GitLab has a "pending" status